### PR TITLE
CE-2897 Adding templates without type to insights

### PR DESCRIPTION
**Pull request by kamilkoterba**
https://wikia-inc.atlassian.net/browse/CE-2897

@Wikia/community-engineering

### CODE REVIEW

File: extensions/wikia/Insights/specials/TemplatesWithoutTypePage.class.php [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+					$title->getDBkey()
+				];
+
+		}
+
+		return $templatesWithoutType;
+	}
+
+	private function getContentNamespacesTemplates($db, $contentNamespaces) {
+		$sql = ( new \WikiaSQL() )
```
Comment (adamkarminski) [inline extensions/wikia/Insights/specials/TemplatesWithoutTypePage.class.php]: `EXPLAIN` yourself :)

**Comment by kamilkoterba** [comment]
@SQLReview

SQL QUERY

```
SELECT DISTINCT p2.page_id as temp_id
     FROM page AS p
         INNER JOIN templatelinks AS t
         ON
         t.tl_from =  p.page_id
         INNER JOIN page AS p2
         ON
         p2.page_title =  t.tl_title
     WHERE p.page_namespace = '0'
     AND p2.page_namespace = '10'
     AND p.page_id <> '1461'
```

EXPLAIN

```
+----+-------------+-------+--------+---------------------------------------------------------------------+------------+---------+--------------------------+-------+-------------------------------------------+
| id | select_type | table | type   | possible_keys                                                       | key        | key_len | ref                      | rows  | Extra                                     |
+----+-------------+-------+--------+---------------------------------------------------------------------+------------+---------+--------------------------+-------+-------------------------------------------+
|  1 | SIMPLE      | p     | ref    | PRIMARY,name_title                                                  | name_title | 4       | const                    | 30594 | Using where; Using index; Using temporary |
|  1 | SIMPLE      | t     | ref    | tl_from                                                             | tl_from    | 4       | gtawiki.p.page_id        |     1 | Using index                               |
|  1 | SIMPLE      | p2    | eq_ref | PRIMARY,name_title,page_random,page_len,page_redirect_namespace_len | name_title | 261     | const,gtawiki.t.tl_title |     1 | Using index                               |
+----+-------------+-------+--------+---------------------------------------------------------------------+------------+---------+--------------------------+-------+-------------------------------------------+
```

File: extensions/wikia/Insights/specials/TemplatesWithoutTypePage.class.php [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+	 * @param bool $offset Only for consistency
+	 * @return array
+	 */
+	public function reallyDoQuery( $limit = false, $offset = false ) {
+		global $wgCityId, $wgContentNamespaces;
+		$dbr = wfGetDB( DB_SLAVE, [ $this->getName(), __METHOD__, 'vslow' ] );
+		$templatesWithoutType = [];
+
+		$CNTemplatesOnWikia = $this->getContentNamespacesTemplates( $dbr, $wgContentNamespaces );
+		$recognizedTemplates  = $this->getNotRecognizedTemplatesOnWikia(
```
Comment (adamkarminski) [inline extensions/wikia/Insights/specials/TemplatesWithoutTypePage.class.php]: Recognized or not recognized, this is a question! :)

File: extensions/wikia/Insights/specials/TemplatesWithoutTypePage.class.php (hunk 1 above)
Comment (adamkarminski) [inline extensions/wikia/Insights/specials/TemplatesWithoutTypePage.class.php]: Actually... Why are you doing this? The acceptance criteria say nothing about focusing only on the Templates included in the content namespace.

File: extensions/wikia/Insights/specials/TemplatesWithoutTypePage.class.php [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+		foreach ( $notRecognizedTemplates as $pageId => $type ) {
+			if ( in_array( $pageId, $cnTemplatesOnWikia ) ) {
+				$cnNotRecognizedTemplates[] = $pageId;
+			}
+		}
+		return $cnNotRecognizedTemplates;
+	}
+
+	private function isRecognized( $type ) {
+		return $type !== TemplateClassificationService::TEMPLATE_UNKNOWN
```
Comment (adamkarminski) [inline extensions/wikia/Insights/specials/TemplatesWithoutTypePage.class.php]: What about creating a static array of these types in one of the more accessible classes and then changing the check to `!in_array( $type, someClass::$unrecognizedTypes )`? Or even moving the whole method there?

**Comment by adamkarminski** [comment]
Besides the comments I left it LGTM. Please, merge the messages that are in dev right now since they were reviewed by Trevor. Thanks!

File: extensions/wikia/Insights/specials/TemplatesWithoutTypePage.class.php [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+		$notRecognizedTemplates  = $this->getNotRecognizedTemplatesOnWikia(
+			( new TemplateClassificationService() ),
+			$wgCityId
+		);
+
+		$cnNotRecognizedTemplates = $this->intersectSets( $cnTemplatesOnWikia, $notRecognizedTemplates );
+
+		foreach( $cnNotRecognizedTemplates as $notRecognizedTemplate ) {
+			$title = Title::newFromID( $notRecognizedTemplate );
+			$links = $title->getIndirectLinks();
```
Comment (adamkarminski) [inline extensions/wikia/Insights/specials/TemplatesWithoutTypePage.class.php]: Just in case - `if ( $title instanceof Title )`? :)

**Comment by sebastianmarzjan** [comment]
Please check (and strongly suggested - remove) code duplication with
https://github.com/Wikia/app/blob/dev/maintenance/wikia/TemplateClassificationRecognizedMetric.php
as those changes look oddly familiar

**Comment by wladekb** [comment]
sqlreview: lgtm

File: includes/wikia/parser/templatetypes/RecognizedTemplatesProvider.class.php [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+		$notRecognizedTemplates = $this->getTemplates( false );
+		if ( empty( $this->namespaces ) ) {
+			return $notRecognizedTemplates;
+		}
+		return $this->intersectSets( $this->getNamespacesTemplates(), $notRecognizedTemplates );
+	}
+
+	public static function isRecognized( $type ) {
+		return $type !== TemplateClassificationService::TEMPLATE_UNKNOWN
+		&& $type !== AutomaticTemplateTypes::TEMPLATE_UNCLASSIFIED
```
Comment (adamkarminski) [inline includes/wikia/parser/templatetypes/RecognizedTemplatesProvider.class.php]: Indentation

File: includes/wikia/parser/templatetypes/RecognizedTemplatesProvider.class.php [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+class RecognizedTemplatesProvider {
+	const ERROR_MESSAGE = 'ExternalTemplatesProviderError';
+
+	private $tcs;
+	private $wikiaId;
+	private $namespaces;
+	private $namespacesTemplates;
+
+	public function __construct( TemplateClassificationService $tcs, $wikiaId, $namespaces = [] ) {
```
Comment (adamkarminski) [inline includes/wikia/parser/templatetypes/RecognizedTemplatesProvider.class.php]: I think that we shouldn't let the marketing language creep into engineering choices and make it `$wikiId` :)

File: includes/wikia/parser/templatetypes/RecognizedTemplatesProvider.class.php [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+	public function getNotRecognizedTemplates() {
+		$notRecognizedTemplates = $this->getTemplates( false );
+		if ( empty( $this->namespaces ) ) {
+			return $notRecognizedTemplates;
+		}
+		return $this->intersectSets( $this->getNamespacesTemplates(), $notRecognizedTemplates );
+	}
+
+	public static function isRecognized( $type ) {
+		return $type !== TemplateClassificationService::TEMPLATE_UNKNOWN
```
Comment (adamkarminski) [inline includes/wikia/parser/templatetypes/RecognizedTemplatesProvider.class.php]: I'd strongly recommend `!in_array()`

File: includes/wikia/parser/templatetypes/RecognizedTemplatesProvider.class.php [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+
+		return $this->namespacesTemplates;
+	}
+
+	/**
+	 * Remove templates from second set that doesn't exist in first set
+	 */
+	private function intersectSets( $namespacesTemplates, $recognizedTemplates ) {
+		$namespacesRecognizedTemplates = [];
+		foreach ( $recognizedTemplates as $pageId => $type ) {
```
Comment (adamkarminski) [inline includes/wikia/parser/templatetypes/RecognizedTemplatesProvider.class.php]: I think that a far better solution to this problem is to construct the `$namespacesTemplates` array in a way that a page id is also a key, so e.g.:

```
[
    31 => 31,
    43 => 43,
]
```

and then using [array_intersect_key](http://php.net/manual/en/function.array-intersect-key.php) to retrieve the common elements.

File: extensions/wikia/Insights/specials/TemplatesWithoutTypePage.class.php [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+		global $wgCityId, $wgContentNamespaces;
+		$templatesWithoutType = [];
+
+		$recognizedProvider = new RecognizedTemplatesProvider(
+			( new TemplateClassificationService() ),
+			$wgCityId,
+			$wgContentNamespaces
+		);
+
+		$cnNotRecognizedTemplates =
```
Comment (adamkarminski) [inline extensions/wikia/Insights/specials/TemplatesWithoutTypePage.class.php]: What happened here? :)

**Comment by adamkarminski** [comment]
Besides the `in_array` and `array_intersect_key` comments the rest is minor, and LGTM. :)

**Comment by kamilkoterba** [comment]
LGTM

**Pull request merged**